runtime/src/iree/hal/utils/deferred_work_queue.c

```diff
@@ -812,6 +820,236 @@ iree_status_t iree_hal_deferred_work_queue_enqueue(
+  action_ptr += wait_semaphore_list_size;
+
+  // Copy signal list for later access.
+  action->signal_semaphore_list.count = signal_semaphore_list.count;
+  action->signal_semaphore_list.semaphores = (iree_hal_semaphore_t**)action_ptr;
+  memcpy(
+      action->signal_semaphore_list.semaphores,
+      signal_semaphore_list.semaphores,
+      signal_semaphore_list.count * sizeof(*signal_semaphore_list.semaphores));
+  action->signal_semaphore_list.payload_values =
+      (uint64_t*)(action_ptr + signal_semaphore_list.count *
+                                   sizeof(*signal_semaphore_list.semaphores));
+  memcpy(action->signal_semaphore_list.payload_values,
+         signal_semaphore_list.payload_values,
+         signal_semaphore_list.count *
+             sizeof(*signal_semaphore_list.payload_values));
+  action_ptr += signal_semaphore_list_size;
+
+  // Copy the execution resources for later access.
+  action->payload.alloc.buffer = buffer;
+
+  // Retain all command buffers and semaphores.
+  iree_status_t status = iree_hal_resource_set_allocate(actions->block_pool,
+                                                        &action->resource_set);
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_resource_set_insert(action->resource_set,
+                                          wait_semaphore_list.count,
+                                          wait_semaphore_list.semaphores);
+  }
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_resource_set_insert(action->resource_set,
+                                          signal_semaphore_list.count,
+                                          signal_semaphore_list.semaphores);
+  }
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_resource_set_insert(action->resource_set, 1, &buffer);
+  }
+
+  if (iree_status_is_ok(status)) {
+    // Now everything is okay and we can enqueue the action.
+    iree_slim_mutex_lock(&actions->action_mutex);
+    if (actions->exit_requested) {
+      status = iree_make_status(
+          IREE_STATUS_ABORTED,
+          "can not issue more executions, exit already requested");
+      iree_hal_deferred_work_queue_action_fail_locked(action, status);
+    } else {
+      iree_hal_deferred_work_queue_action_list_push_back(&actions->action_list,
+                                                         action);
+      // One work item is the callback that makes it across from the
+      // completion thread.
+      actions->pending_work_items_count += 1;
+    }
+    iree_slim_mutex_unlock(&actions->action_mutex);
+  } else {
+    iree_hal_resource_set_free(action->resource_set);
+    iree_allocator_free(actions->host_allocator, action);
+  }
+
+  IREE_TRACE_ZONE_END(z0);
+  return status;
+}
+
+// Enqueues command buffer submissions into the work queue to be executed
+// once all semaphores have been satisfied.
+iree_status_t iree_hal_deferred_work_queue_enqueue_dealloc(
```

Comment:
I think this function is identical to the one above just with IREE_HAL_QUEUE_ACTION_TYPE_QUEUE_DEALLOC?
you can make a `iree_hal_deferred_work_queue_enqueue_buffer_operation(type)` impl that you call from both